templating/secrets.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/pkg/errors"
+)
+
+// ResolveSecretFunc is a function that resolves the secret to its value and sends through the ResolvedChan of the secret. Any errors during resolve are send through errorChan
+type ResolveSecretFunc func(ctx context.Context, azureVaultResourceURL string, secret *graph.Secret, errorChan chan error)
+
+// SecretResolver defines how a secret is resolved.
+type SecretResolver struct {
+	Resolve          ResolveSecretFunc
+	azureEnvironment azure.Environment
+}
+
+// NewSecretResolver creates a resolver with the given resolve function.
+func NewSecretResolver(resolveFunc ResolveSecretFunc, azureEnvironmentName string) (*SecretResolver, error) {
+	if azureEnvironmentName == "" {
+		azureEnvironmentName = azure.PublicCloud.Name
+	}
+
+	azureEnvironment, err := azure.EnvironmentFromName(azureEnvironmentName)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid azure environment name")
+	}
+
+	return &SecretResolver{Resolve: resolveFunc, azureEnvironment: azureEnvironment}, nil
+}
+
+// DefaultSecretResolver creates a default secret resolver for the given azure environment
+func DefaultSecretResolver(azureEnvironmentName string) (*SecretResolver, error) {
+	return NewSecretResolver(resolveSecret, azureEnvironmentName)
+}
+
+// ResolveSecrets returns a list of resolved secrets and returns error if there is any failure in resolving a secret.
+func (secretResolver *SecretResolver) ResolveSecrets(ctx context.Context, secrets []*graph.Secret) (Values, error) {
+
+	resolvedSecrets := Values{}
+
+	if len(secrets) == 0 {
+		return resolvedSecrets, nil
+	}
+
+	// We will resolve in batches of 5 to avoid throttling errors on the vault providers
+	batchSize := 5
+	errorChan := make(chan error)
+	for index := 0; index < len(secrets); index += batchSize {
+		endIndex := index + batchSize
+
+		if endIndex > len(secrets) {
+			endIndex = len(secrets)
+		}
+
+		var resolvedChans []chan graph.SecretValue
+
+		for _, secret := range secrets[index:endIndex] {
+			// Validate secret before resolving
+			err := secret.Validate()
+			if err != nil {
+				if secret.ID == "" {
+					return resolvedSecrets, err
+				}
+				return resolvedSecrets, errors.Wrap(err, fmt.Sprintf("failed to validate secret with ID: %s", secret.ID))
+			}
+
+			if secret == nil {
+				continue
+			}
+
+			if secret.ResolvedChan == nil {
+				secret.ResolvedChan = make(chan graph.SecretValue)
+			}
+			resolvedChans = append(resolvedChans, secret.ResolvedChan)
+			go secretResolver.Resolve(ctx, secretResolver.azureEnvironment.KeyVaultEndpoint, secret, errorChan)
+		}
+
+		// Block until either:
+		// - The global context expires
+		// - Resolving a secret has error
+		// - All secrets are resolved successfully
+		for _, ch := range resolvedChans {
+			select {
+			case <-ctx.Done():
+				return resolvedSecrets, ctx.Err()
+			case secretValue := <-ch:
+				resolvedSecrets[secretValue.ID] = secretValue.Value
```

Comment:
How does the code handle two secrets with the same id? Will it fail somewhere in the pipeline or just overwrite?

Reply:
yea, race for the overwrite - maybe you can add validation during task creation for it

Reply:
My understanding was the last write will be success just like we override the values? If not I can fail the task if it has duplicate IDs

Reply:
personally I'm ok with leaving it as-is, maybe make an issue to track it

Reply:
Override is fine as long as the result is predictable and consistent. Since the secret resolver is not executed in order, the result may not be consistent.